feat: LSDV-4915: Async import

triklozoid opened the pull request.

triklozoid committed:
feat: LSDV-4915: Async import

triklozoid committed:
Fix tests (LSDV-4915)

triklozoid committed:
Add docs; fix all_urls

triklozoid committed:
Fix import from url (LSDV-4915)

triklozoid committed:
Merge branch 'develop' into fb-LSDV-4915

triklozoid committed:
Add error field (LSDV-4915)

bmartel commented on label_studio/data_import/uploader.py:
```diff
@@ -145,33 +145,95 @@ def tasks_from_url(file_upload_ids, project, request, url, could_be_tasks_lists)
+        # try to load json with task or tasks from url as string
+        json_data = str_to_json(url)
+        if json_data:
+            file_upload = create_file_upload(request.user, project_import.project, SimpleUploadedFile('inplace.json', url.encode()))
```
```suggestion
            file_upload = create_file_upload(user, project_import.project, SimpleUploadedFile('inplace.json', url.encode()))
```

request is no longer provided in-context, need to update this to user.

triklozoid committed:
Fix import from url

Co-authored-by: bmartel <brandonmartel@gmail.com>

makseq commented on label_studio/core/redis.py:
```diff
@@ -110,6 +110,7 @@ def start_job_async_or_sync(job, *args, in_seconds=0, **kwargs):
         )
         return job
     else:
+        kwargs.pop('on_failure', None)
```
is it ok to pop it here?

triklozoid replied:
Generally, it's ok, I rewrote it for proper error handling.

makseq commented on label_studio/data_import/functions.py:
```diff
@@ -0,0 +1,107 @@
+logger = logging.getLogger(__name__)
+
+
+def async_import_background(import_id, user_id):
```
async_import_background(import_id, user_id, **kwargs):

and pass project_id and organization_id there, it will help a lot for /django-rq page and other debugs.

triklozoid replied:
Fixed.

makseq commented on label_studio/data_import/functions.py:
```diff
@@ -0,0 +1,107 @@
+        prediction_count = len(serializer.db_predictions)
+        # Update counters (like total_annotations) for new tasks and after bulk update tasks stats. It should be a
+        # single operation as counters affect bulk is_labeled update
+        project.update_tasks_counters_and_task_states(tasks_queryset=tasks, maximum_annotations_changed=False,
```
let's add new line after `(`, like black linter would do it

triklozoid replied:
Fixed. 
@makseq Do we plan to add the black linter in this repo?

makseq replied:
yeh, I would add it, but we can't process all files at once, because we will get lots of merge conflicts. So, I would set up black in this repo, but I would add django app one by one to it.

makseq commented on label_studio/projects/api.py:
```diff
@@ -368,6 +368,26 @@ def get(self, *args, **kwargs):
+class ProjectImportAPI(generics.RetrieveAPIView):
+    parser_classes = (JSONParser,)
+    serializer_class = ProjectImportSerializer
+    permission_required = all_permissions.projects_view
```
projects_change?

triklozoid replied:
Fixed.

makseq commented on label_studio/data_import/uploader.py:
```diff
@@ -123,7 +123,7 @@ def str_to_json(data):
 
 
-def tasks_from_url(file_upload_ids, project, request, url, could_be_tasks_lists):
+def tasks_from_url(file_upload_ids, project, user, url, could_be_task_list):
```
I would rename it as could_be_task**s**_list, because ProjectImport has this field already. In this case we will have at least 2 variants instead of 3 (could_be_task_list, could_be_tasks_list, could_be_tasks_lists)

triklozoid replied:
Fixed.

makseq commented:
**I tried to run it in OS.**

1. **(middle)** `fflag_feat_all_lsdv_4915_async_task_import_13042023_short=true` and tasks are not imported:
https://user-images.githubusercontent.com/501780/234166459-7e9d2362-f7dc-4e7b-99c0-8c03791b3b83.mp4

but seems we have to keep this FF is disabled for opensource always? if yes, it's kind of a problem... better to add some check for community version in the code there along with a feature flag check:

```
 if flag_set('fflag_feat_all_lsdv_4915_async_task_import_13042023_short', request.user):
    return self.async_import(request, project, preannotated_from_fields, commit_to_project, return_task_ids)
else:
    return self.sync_import(request, project, preannotated_from_fields, commit_to_project, return_task_ids)
```

When `fflag_feat_all_lsdv_4915_async_task_import_13042023_short=false`, import works well.

makseq commented:
**I tried to run it in ENT.**


2. **(high)** import csv / txt has a regression, compare these two screens (first - async import, second - app.heartex.com):

![image](https://user-images.githubusercontent.com/501780/234170929-f17eae48-c5d2-4faa-948c-2b5c1bd44a91.png)

![image](https://user-images.githubusercontent.com/501780/234170972-86c03510-2239-49c5-9ae9-aa5b40af9431.png)


4. **(low)** if import fails, it will never finish (progress bar is infinite) and no errors will be displayed.

(It's not a simple to solve. In storage StatusInfo I introduced time_last_ping variable and it is updated every 10 seconds, based on it + based on rqworker job state I know the actual storage status). However, maybe for this first async implementation of the import it's enough, user can simply click Cancel and retry importing again. 


5. **(highest)** The most loaded part of import still works synchronously. 

![image](https://user-images.githubusercontent.com/501780/234172726-0b0d9cda-7e25-4651-858c-2106576686da.png)

This breakpoint never worked for me, it's always executed as a synchronous call in ImportAPI._save() when I click `Import` button after the file is uploaded:

![image](https://user-images.githubusercontent.com/501780/234173091-e1979178-b630-438d-8ae3-6b99fd56fa7e.png)

![image](https://user-images.githubusercontent.com/501780/234173034-2c488185-8bf2-4983-9423-2f09a1a0cac3.png)

triklozoid committed:
Merge branch 'develop' into fb-LSDV-4915

triklozoid committed:
Review fixes(LSDV-4915)

triklozoid committed:
Fix merge(LSDV-4915)

triklozoid committed:
Fix could_be_tasks_list detection (LSDV-4915)

triklozoid commented:
@makseq 

1. Added check for `Community`
2. Fixed, should work now. 
4. Left unchanged, for first implementation I think it's okay. 
5. `reimport` request is out of scope of this task, it will be fixed in LSDV-4971

triklozoid committed:
Fix on_failure (LSDV-4915)

triklozoid committed:
Fix typo in docs (LSDV-4915)

triklozoid committed:
Add debug logging for file upload (LSDV-4915)

triklozoid committed:
Merge branch 'develop' into fb-LSDV-4915
